File: src/main/java/com/naminhyeok/fantazzk/template/application/CreateTemplate.java
package com.naminhyeok.fantazzk.template.application;

import com.naminhyeok.fantazzk.template.domain.Template;
import com.naminhyeok.fantazzk.template.domain.TemplatePlayer;
import com.naminhyeok.fantazzk.template.repository.Templates;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class CreateTemplate {
    private final Templates templates;

    @Transactional
    public Template create(CreateTemplateCommand command) {
        Template template = switch (command) {
            case CreateTemplateCommand.Auction auction ->
                Template.createAuction(
                    auction.name(),
                    auction.gameType(),
                    auction.teamCount(),
                    auction.teamSize(),
                    auction.budget(),
                    auction.pickBanTime(),
                    auction.minBidUnit(),
                    auction.positionLimit(),
                    auction.players().stream()
                        .map(player -> new TemplatePlayer(player.name(), player.position(), player.displayOrder()))
                        .toList()
                );
            case CreateTemplateCommand.Draft draft ->
                Template.createDraft(
                    draft.name(),
                    draft.gameType(),
                    draft.teamCount(),
                    draft.teamSize(),
                    draft.pickBanTime(),
                    draft.strategy(),
                    draft.players().stream()
                        .map(player -> new TemplatePlayer(player.name(), player.position(), player.displayOrder()))
                        .toList()
                );
        };

        return templates.save(template);
    }
}

File: src/main/kotlin/com/naminhyeok/fantazzk/template/application/CreateTemplate.kt
package com.naminhyeok.fantazzk.template.application

import com.naminhyeok.fantazzk.template.domain.DraftOrderStrategy
import com.naminhyeok.fantazzk.template.domain.Template
import com.naminhyeok.fantazzk.template.repository.Templates
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

sealed interface CreateTemplateCommand {
    val name: String
    val teamCount: Int
    val teamSize: Int
    val playerNames: List<String>

    data class Auction(
        override val name: String,
        override val teamCount: Int,
        override val teamSize: Int,
        val budget: Int,
        override val playerNames: List<String>,
    ) : CreateTemplateCommand

    data class Draft(
        override val name: String,
        override val teamCount: Int,
        override val teamSize: Int,
        val strategy: DraftOrderStrategy,
        override val playerNames: List<String>,
    ) : CreateTemplateCommand
}

@Component
class CreateTemplate(
    private val templateRepository: Templates,
) {
    @Transactional
    fun create(command: CreateTemplateCommand): Template {
        val template =
            when (command) {
                is CreateTemplateCommand.Auction ->
                    Template.createAuction(command.name, command.teamCount, command.teamSize, command.budget, command.playerNames)

                is CreateTemplateCommand.Draft ->
                    Template.createDraft(command.name, command.teamCount, command.teamSize, command.strategy, command.playerNames)
            }
        return templateRepository.save(template)
    }
}
